### PR TITLE
Update build to use a local package cache from a submodule and build offline

### LINKED ISSUE #53710: Handle 3rd party NPM dependencies in source-build
We recently attempted to turn on the build of the NPM components in source-build. This led to issues as the NPM/node components have a couple 3rd party dependencies, which can't be restored in offline source-build. We have a couple options for how to tackle this:

1. Check in a node package cache (matching our package.lock.json file) to a separate repo, which is a submodule of aspnetcore. Any time a dev changes a node dependency, they should re-generate the cache, push it to the new repo, and update the submodule in their PR. We'll need  to exclude that repo from automated submodule updates in that case (dependabot). Node has a validation mode that we can run during the build to ensure the package cache matches what's in package.lock.json.

2. Add a pre-step for source build that does an online restore before we start offline source-build. There are already online pre-steps for source-build, so this wouldn't be against the "spirit" of source-build.

We should choose whichever approach most closely matches what our source-build partners do.

CC @dotnet/distro-maintainers @javiercn @mthalman @MichaelSimons

### PR DESCRIPTION
**Pull request by wtgodbe**
Replaces https://github.com/dotnet/aspnetcore/pull/53524/files to fix a casing issue with the submodule. Should fix https://github.com/dotnet/aspnetcore/issues/53710.

### CODE REVIEW

File: eng/Npm.Workspace.nodeproj [hunk 1]
```diff
@@ -15,9 +15,14 @@
     <PackageVersion>$(VersionPrefix)$(VersionSuffix)</PackageVersion>
   </PropertyGroup>
 
-  <Target Name="Restore">
+  <Target Name="_VerifyNPMCache">
```
Comment (wtgodbe) [inline eng/Npm.Workspace.nodeproj:18]: @javiercn is this the right time to call `npm cache verify` (before Restore)?

**Review by mthalman — APPROVED** [review]

**Comment by MichaelSimons** [comment]
Wanted to get your thoughts on if this warrants validating in the VMR prior to merging? I would like to avoid surprises, regressing SB CI, and/or blocking the sdk->installer dependency flow.

**Comment by wtgodbe** [comment]
> Wanted to get your thoughts on if this warrants validating in the VMR prior to merging? I would like to avoid surprises, regressing SB CI, and/or blocking the sdk->installer dependency flow.

I think that's a good idea. How can I help get that done?

File: eng/Npm.Workspace.nodeproj (hunk 1 above)
Comment (javiercn) [reply eng/Npm.Workspace.nodeproj:18]: Yep

**Review by javiercn — APPROVED** [review]

**Review by javiercn — COMMENTED** [review]
We also need to remove the checked in JS files, don't we?

**Comment by wtgodbe** [comment]
> We also need to remove the checked in JS files, don't we?

Good point, doing that now

**Comment by javiercn** [comment]
@wtgodbe I'm merging this as I'm doing npm dependency updates as part of build ops and I don't want to create conflicts for you.

**Comment by MichaelSimons** [comment]
> > Wanted to get your thoughts on if this warrants validating in the VMR prior to merging? I would like to avoid surprises, regressing SB CI, and/or blocking the sdk->installer dependency flow.
> 
> I think that's a good idea. How can I help get that done?

I see this is already merged but for future reference, you can either create a git patch with your changes and apply it to the [vmr](https://github.com/dotnet/dotnet) or use this [vmr-sync script](https://github.com/dotnet/installer/blob/main/eng/vmr-sync.sh).  In this particular case involving the submodule, the vmr-sync would be the recommended approach.  Once the changes are applied, I would recommend just opening a draft PR and let the CI run to validate.  If you would like to build locally, the [build instructions](https://github.com/dotnet/dotnet?tab=readme-ov-file#building) are in the VMR's readme

**Comment by MichaelSimons** [comment]
It would still be good to validate this IMO to ensure this does not block the sdk->installer dependency flow or regress source-build.

**Comment by MichaelSimons** [comment]
> It would still be good to validate this IMO to ensure this does not block the sdk->installer dependency flow or regress source-build.

On another look, this won't affect source-build because BuildNodeJS is not set to true in SB.  Is enabling this in source-build on someone's radar/tracked in an issue?

**Comment by mthalman** [comment]
This seems to be causing a failure: https://github.com/dotnet/installer/pull/18641#issuecomment-1942399952

**Comment by wtgodbe** [comment]
Responded here: https://github.com/dotnet/installer/pull/18641#issuecomment-1942813124

**Comment by wtgodbe** [comment]
> On another look, this won't affect source-build because BuildNodeJS is not set to true in SB. Is enabling this in source-build on someone's radar/tracked in an issue?

It's already enabled in this repo's source-build leg. I'm enabling it in Installer in https://github.com/dotnet/installer/pull/18641

**Comment by MichaelSimons** [comment]
This appears to have regressed all of the offline source-build legs in CI - https://github.com/dotnet/source-build/issues/4129

**Comment by javiercn** [comment]
@MichaelSimons The error seems to be caused by a component that is likely not needed and that we can avoid using. Is there a way to setup an environment variable in source build? 

We probably just need to set `PUPPETEER_SKIP_DOWNLOAD=1` but I don't know if that's done here or in the source build config/definition.

**Comment by javiercn** [comment]
https://github.com/dotnet/aspnetcore/pull/54036

**Pull request merged**